pkg/utils/k8sutil.go

```diff
@@ -543,6 +543,202 @@ func EnsureFuncService(client kubernetes.Interface, funcObj *kubelessApi.Functio
+	if len(tag) < 64 {
+		return fmt.Errorf("Expecting sha256 as image tag")
+	}
+	jobName := fmt.Sprintf("build-%s-%s", funcObj.ObjectMeta.Name, tag[0:10])
+	_, err := client.BatchV1().Jobs(funcObj.ObjectMeta.Namespace).Get(jobName, metav1.GetOptions{})
+	if err == nil {
+		// The job already exists
+		logrus.Infof("Found a previous job for building %s:%s", imageName, tag)
+		return nil
+	}
+	podSpec := v1.PodSpec{
+		RestartPolicy: v1.RestartPolicyOnFailure,
+	}
+	runtimeVolumeMount := getRuntimeVolumeMount(funcObj.ObjectMeta.Name)
+	err = populatePodSpec(funcObj, lr, &podSpec, runtimeVolumeMount)
+	if err != nil {
+		return err
+	}
+
+	// Add a final initContainer to create the function bundle.tar
+	prepareContainer := v1.Container{}
+	for _, c := range podSpec.InitContainers {
+		if c.Name == "prepare" {
+			prepareContainer = c
+		}
+	}
+	podSpec.InitContainers = append(podSpec.InitContainers, v1.Container{
+		Name:         "bundle",
+		Command:      []string{"sh", "-c"},
+		Args:         []string{fmt.Sprintf("tar cvf %s/bundle.tar %s/*", runtimeVolumeMount.MountPath, runtimeVolumeMount.MountPath)},
+		VolumeMounts: prepareContainer.VolumeMounts,
+		Image:        unzip,
+	})
+
+	buildJob := batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            jobName,
+			Namespace:       funcObj.ObjectMeta.Namespace,
+			OwnerReferences: or,
+			Labels: map[string]string{
+				"created-by": "kubeless",
+				"function":   funcObj.ObjectMeta.Name,
+			},
+		},
+		Spec: batchv1.JobSpec{
+			Template: v1.PodTemplateSpec{
+				Spec: podSpec,
+			},
+		},
+	}
+
+	baseImage, err := lr.GetFunctionImage(funcObj.Spec.Runtime, funcObj.Spec.Type)
+	if err != nil {
+		return err
+	}
+
+	// Registry volume
+	dockerCredsVol := imagePullSecretName
+	dockerCredsVolMountPath := "/docker"
+	registryCredsVolume := v1.Volume{
+		Name: dockerCredsVol,
+		VolumeSource: v1.VolumeSource{
+			Secret: &v1.SecretVolumeSource{
+				SecretName: imagePullSecretName,
+			},
+		},
+	}
+	buildJob.Spec.Template.Spec.Volumes = append(buildJob.Spec.Template.Spec.Volumes, registryCredsVolume)
+
+	// Add main container
+	buildJob.Spec.Template.Spec.Containers = []v1.Container{
+		{
+			Name:  "build",
+			Image: builderImage,
+			VolumeMounts: append(prepareContainer.VolumeMounts,
+				v1.VolumeMount{
+					Name:      dockerCredsVol,
+					MountPath: dockerCredsVolMountPath,
+				},
+			),
+			Env: []v1.EnvVar{
+				{
+					Name:  "DOCKER_CONFIG_FOLDER",
+					Value: dockerCredsVolMountPath,
+				},
+			},
+			Args: []string{
+				"/imbuilder",
+				"add-layer",
+				"--src", fmt.Sprintf("docker://%s/%s", registryHost, baseImage),
+				"--dst", fmt.Sprintf("docker://%s/%s:%s", registryHost, imageName, tag),
+				fmt.Sprintf("%s/bundle.tar", podSpec.InitContainers[0].VolumeMounts[0].MountPath),
+			},
+		},
+	}
+
+	// Create the job if doesn't exists yet
```

Comment:
Please add a log message with builder job name with function name if job is created.